### ISSUE TITLE
Allow Asn1Writable types to provide their length

### ISSUE THREAD

**Issue by alex** [alex/rust-asn1#340]
Right now we reserve space for the length, call `write()`, see how much data was written, and fill it in (potentially resizing the storage if more than 127 bytes are written). We should allow types to pre-declare their size to avoid resizing.

**Comment by dishmaker** [comment]
You can define fn `encoded_len` in trait, like in the `der` crate:
https://github.com/RustCrypto/formats/blob/f66f4c5f1fd53d9744a0837cf0913585b717d071/der/src/encode.rs#L25

That would remove the need for prepending 0x00 byte.
https://github.com/alex/rust-asn1/blob/2e1a8c48e4ea7c9cf58809868206190700a0aff0/src/writer.rs#L120-L122

**Comment by alex** [comment]
Yes, that's the desired approach here. (I think probably it should be `Option<usize>`, instead of `Result<>`).
